### ISSUE TITLE
Docker containers scope

### ISSUE THREAD

**Issue by sharksforarms** [pytest-docker-compose/pytest-docker-compose#3]
Is there a way to have the docker containers up for multiple tests?

Example

```
@pytest.mark.parametrize("var", [1, 2, 3])
def test_mytest(docker_containers, var):
    print(var)
```

Current
```
<docker compose up>
print(1)
<docker compose down>
<docker compose up>
print(2)
<docker compose down>
<docker compose up>
print(3)
<docker compose down>
```

Desired:
```
<docker compose up>
print(1)
print(2)
print(3)
<docker compose down>
```

I think there might be a way with pytest scopes, but currently the scope is function based

**Comment by todofixthis** [comment]
Hey @sharksforarms!  That's pretty impressive that you can type with sharks for arms; how do you protect your keyboards from getting eaten?? [Maybe you wear boxing gloves](http://hrwiki.org/wiki/How_do_you_type_with_boxing_gloves_on%3F) 😸

Anyway.  The decision to bring containers down after each test was a deliberate one — should a test cause a change in one of the Docker containers that doesn't get cleaned up by the end of that test, it could impact the results of subsequent tests.

That said, it does take a long time, and there might be some cases where it's OK to keep the containers running after a test.  Hm.

As an experiment, try forking this repo and replacing this line:

https://github.com/centralityai/pytest-docker-compose/blob/1.0.1/pytest_docker_compose/__init__.py#L74

Change `@pytest.fixture` to `@pytest.fixture(scope="session")` — that should cause the containers to get torn down only at the end of the test session.

Let me know if that works, and we can look into a more configurable implementation (:

**Comment by sharksforarms** [comment]
Exactly, some tests I need a fresh docker-compose environment but other tests (for speed sake) I can have the same environment since the tests are independent of each other.

By changing it to session, this is now using the same environment for all tests. The class scope DOES seem to work though

`scope=session`
```
class TestClass(object):
    @pytest.mark.parametrize("var", [1, 2, 3])
    def test_mytest(self, docker_containers, var):
        print('Test1', var)
    @pytest.mark.parametrize("var", [1, 2, 3])
    def test_test2(self, docker_containers, var):
        print('Test2', var)

@pytest.mark.parametrize("var", [1, 2, 3])
def test_test3(docker_containers, var):
    print('Test3', var)
```
Output:
```
<docker_compose up>
Test1 1
Test1 2
Test1 3
Test2 1
Test2 2
Test2 3
Test3 1
Test3 2
Test3 3
<docker compose down>
```

`scope=class`
```
class TestClass(object):
    @pytest.mark.parametrize("var", [1, 2, 3])
    def test_mytest(self, docker_containers, var):
        print('Test1', var)
    @pytest.mark.parametrize("var", [1, 2, 3])
    def test_test2(self, docker_containers, var):
        print('Test2', var)

@pytest.mark.parametrize("var", [1, 2, 3])
def test_test3(docker_containers, var):
    print('Test3', var)
```
Output:
```
<docker_compose up>
Test1 1
Test1 2
Test1 3
Test2 1
Test2 2
Test2 3
<docker compose down>
<docker_compose up>
Test3 1
Test3 2
Test3 3
<docker compose down>
```

If I want to  keep the same set of docker-containers up for a set of tests I can wrap it in a class like above

**Comment by todofixthis** [comment]
Interesting.  I like it; I'll do a little research and see if pytest provides an interface for this.

**Comment by mathom** [comment]
I'm also interested in a configurable solution to this! I've had to fork and modify it - I'm not sure if pytest has a good way to change fixture scope dynamically...

**Comment by sharksforarms** [comment]
There's a way to define multiple fixture scopes like so:

```
    def _docker_containers(self, docker_project: Project):
        ...
        yield containers
        ...

    @pytest.fixture
    def docker_containers(self, docker_project: Project):
        yield from self._docker_containers(docker_project)

    @pytest.fixture(scope="class")
    def docker_containers_class(self, docker_project: Project):
        yield from self._docker_containers(docker_project)
```

Not sure what the best practice is

**Comment by thomasleveil** [comment]
maybe `scope="module"` would be convenient to have. It would make it very easy for users to group tests that should use the same containers together in one python module file.

```python
@pytest.fixture(scope="module")
    def docker_containers_module(self, docker_project: Project):
        yield from self._docker_containers(docker_project)
```

**Comment by todofixthis** [comment]
I'm looking into this now.  It seems that changing a fixture's scope dynamically is not supported in pytest (pytest-dev/pytest#1682), and I'm hesitant to implement my own version of that functionality.

My main concern is the possibility of weird issues caused by a fixture's scope being changed mid-scope (e.g., a session-scoped fixture is changed to function scope partway through, or vice versa).  Then, take that complexity and square it because there's two fixtures whose scope would need to be kept in sync (`docker_containers`, `docker_network_info`).

That's a lot of complexity that would require a ton of testing to get right, and unfortunately it's not something that I can dedicate much time to right now ):

What I'll do for now is make it easier to declare your own versions of the `docker_containers` and `docker_network_info` fixtures that you can scope however you want, but they will be marked private, so that it's explicit that this is not covered by the warranty.

PR imminent.  Watch this space.

**Comment by todofixthis** [comment]
An alternative approach to this problem would be to keep the function-scoped fixtures, but find a way to run tests in parallel.

**Comment by todofixthis** [comment]
I just uploaded some changes to my personal repo here:
https://github.com/todofixthis/pytest-docker-compose/tree/feature/3-up-periscope

Here is an example of changing scope to session (based on [docker-compose "Hello World"](https://docs.docker.com/compose/gettingstarted/)):

```python
import typing
from time import sleep
from urllib.request import urlopen

import pytest

from pytest_docker_compose import DockerComposePlugin, NetworkInfo


@pytest.fixture(scope="session", name="session_containers")
def fixture_session_containers(docker_project):
    containers = DockerComposePlugin._containers_up(docker_project)
    yield containers
    DockerComposePlugin._containers_down(docker_project, containers)


@pytest.fixture(scope="session", name="session_network_info")
def fixture_session_network_info(session_containers):
    return DockerComposePlugin._extract_network_info(session_containers)


def test_hello(
        session_network_info: typing.Dict[str, typing.List[NetworkInfo]],
):
    host = session_network_info["docker_web_1"][0]

    attempts = 0
    while True:
        try:
            contents: bytes = (
                urlopen(f"http://{host.hostname}:{host.host_port}/").read()
            )
        except ConnectionError:
            attempts += 1
            assert attempts < 5
            sleep(0.1)
        else:
            break

    assert contents == b"Hello World! I have been seen 1 times.\n"


def test_persistence(
        session_network_info: typing.Dict[str, typing.List[NetworkInfo]],
):
    host = session_network_info["docker_web_1"][0]

    attempts = 0
    while True:
        try:
            contents: bytes = (
                urlopen(f"http://{host.hostname}:{host.host_port}/").read()
            )
        except ConnectionError:
            attempts += 1
            assert attempts < 5
            sleep(0.1)
        else:
            break

    assert contents == b"Hello World! I have been seen 2 times.\n"
```

**Comment by todofixthis** [comment]
Hey @sharksforarms @vorstelladuffy @thomasleveil were you able to try out the patched version in your environment?

**Comment by sharksforarms** [comment]
I haven't had the chance to try it, but it looks good to me!

**Comment by vorstelladuffy** [comment]
@todofixthis - I just got a chance to revisit this project. Your branch fork works well for me! Thanks.

**Comment by todofixthis** [comment]
Awesome!  Thanks @vorstelladuffy!!

**Comment by RmStorm** [comment]
> An alternative approach to this problem would be to keep the function-scoped fixtures, but find a way to run tests in parallel (e.g., using [pytest-xdist](https://github.com/pytest-dev/pytest-xdist)).
> 
> Tricky bit would be managing multiple sets of docker containers — maybe there's a way to inject a prefix into network/service names, and/or run each subprocess on a separate instance/container.

I gave a try to implementing this but I figured out there is an issue with this approach. Renaming the default network and setting up multiple networks is not a big problem but the services are usually exposed on localhost!

When you create multiple networks that have an identical configuration except for the network name you will run into conflicts since the services try to expose themselves on the same port on localhost. It is off course possible to work around this as well but that will get real fragile real quick.

**Comment by richardhanson** [comment]
@RmStorm took a look at your enhancements. Like it a lot! Forked your version to make my own updates to address #8. Glad to see tests there as well. Would love to see this all merged back into pytest-docker-compose and uploaded to PyPI (@todofixthis).

**Comment by todofixthis** [comment]
100% agree 😺  I no longer work at Centrality, so unfortunately I am no longer able to maintain this project.  However, I am still in contact with folks there, and I've asked them to have a look 🙂

**Comment by sharksforarms** [comment]
Hey @todofixthis any updates? Is there currently someone maintaining the project?

**Comment by RmStorm** [comment]
> Hey @todofixthis any updates? Is there currently someone maintaining the project?

I don't think there is a lot of active maintenance going on. I made a PR for a small feature a couple months ago and it took two weeks to merge. The beauty of open source is off course that this is not a big problem, you can just fork.

I have made a fork with the feature you need, e.g. scoping containers to 'function', 'class', 'module' or 'session'. I'm publishing on a private repository on gemfury to avoid confusion on pypi so if you want to use it you need to fork my repo from [here](https://github.com/RmStorm/pytest-docker-compose/tree/gemfury-master).

However as there isn't really anyone maintaining the project currently I would like to take it over and publish a newer version pypi. It would also be nice to merge @richardhanson his work from #8 in.

@todofixthis @maticko is there any chance of that being a possibility?

**Comment by richardhanson** [comment]
I was actually considering the same thing myself @RmStorm and would be happy to help you maintain the package if it gets handed over. If nobody responds in the next month or so, I would suggest that we just release under a different name on PyPI.

**Comment by todofixthis** [comment]
Love the initiative here.  You've put a tonne of work into your fork, and I would hate for the broader development community to be deprived of a quality plugin because of admin overhead.

If you decide to release under a different name on PyPI, I will support that decision 👍🏻

**Comment by maticko** [comment]
As a current owner of this repo, what do you guys suggest as a solution? We don't have any plan to maintain this. I can ask the business if they are willing to transfer this repository to someone else.

**Comment by richardhanson** [comment]
I think the ideal solution is transferring the repo to @RmStorm as well as transferring the package on PyPI. The PyPI part is required for the package to retain its name, the repo is a nice to have.

**Comment by sharksforarms** [comment]
I support this! Awesome initiative indeed.

**Comment by RmStorm** [comment]
I would definitely be up for mainlining the repo, I am currently actively using this on several projects and will be for the foreseeable future.

So if @maticko gets a green light to transfer we can proceed. I have made an account on PyPI as roaldstorm@gmail.com so transferring ownership there should be straightforward. For the repository I agree with @richardhanson it would be nice to have.

What do you guys think is best? Should I create an empty organization for it?

**Comment by richardhanson** [comment]
@RmStorm I think that would be best. In case you no longer want to support the library, others can easily take over instead of needing to transfer the repo again.

**Comment by maticko** [comment]
Hey guys, just an update. I didn't get the final yes from the business yet. Will keep chasing the guys.

**Comment by RmStorm** [comment]
> 
> 
> Hey guys, just an update. I didn't get the final yes from the business yet. Will keep chasing the guys.

Great! As soon as you get the go-ahead I will create an organization for the repo and make you owner as well. I think that's required to transfer ownership.

**Comment by maticko** [comment]
@RmStorm I got confirmation. Let's transfer this.

**Comment by RmStorm** [comment]
I've made an organization and send you a member invite with ownership role. Having ownership in the organization should allow you to trasnfer the repo from centralityai to https://github.com/orgs/pytest-docker-compose

**Comment by RmStorm** [comment]
Version 2.0.0 is just published please read the new [README](https://github.com/pytest-docker-compose/pytest-docker-compose/blob/develop/README.rst) for an explanation of how to use the new version.

There is new fixtures called 'function_scoped_containers', 'class_scoped_containers', 'module_scoped_containers' and 'session_scoped_containers'.

These fixtures bring up the docker compose project for a certain scope. It is off course important to use them properly. For example if you have a test_my_containers.py file and you use both function and module scoped container you will receive an error. This is because the module scoped containers session will bring up the containers and it will only bring them down when all tests in that module have completed. If you are trying to run a test that than uses function scoped containers it too will try to bring up the containers which is not possible since they are already running.

Anyhow, like I said. Read the new docs =D they will help you get started. You can also take a look at the new testing code for examples of how to use the plugin.

**Comment by nicoddemus** [comment]
> I'm looking into this now. It seems that changing a fixture's scope dynamically is not supported in pytest ([pytest-dev/pytest#1682](https://github.com/pytest-dev/pytest/issues/1682)), and I'm hesitant to implement my own version of that functionality.

Hey folks!

I came here through https://github.com/pytest-dev/pytest/pull/5776: so the good news is that @aklajnert volunteered to implement dynamic scopes. 🎉 

We could use some help though, as we are struggling to come up with a good example to put in the docs, so if someone could help to provide a good example with your use case it would be neat!

https://github.com/pytest-dev/pytest/pull/5776/files#diff-ed6ef9c09ea724ad4bab816b6a753db7

**Comment by RmStorm** [comment]
> > I'm looking into this now. It seems that changing a fixture's scope dynamically is not supported in pytest ([pytest-dev/pytest#1682](https://github.com/pytest-dev/pytest/issues/1682)), and I'm hesitant to implement my own version of that functionality.
> 
> Hey folks!
> 
> I came here through [pytest-dev/pytest#5776](https://github.com/pytest-dev/pytest/pull/5776): so the good news is that @aklajnert volunteered to implement dynamic scopes. tada
> 
> We could use some help though, as we are struggling to come up with a good example to put in the docs, so if someone could help to provide a good example with your use case it would be neat!
> 
> https://github.com/pytest-dev/pytest/pull/5776/files#diff-ed6ef9c09ea724ad4bab816b6a753db7

Hey @nicoddemus that's really cool! I think pytest-docker-compose is a really good example of when you would need dynamic scoping. Basically we have a fixture that spins up a bunch of docker containers and than let's you wait for them to become responsive. After that you can interact with the started containers.

However if there are many containers you really want to start all of them just once and then keep them going for the duration of all tests (session scope). If you have just a few containers it can make sense to start and stop them so that you have a clean slate for every test (function scope) or file (module scope).

So the way I currently implemented this is a little hacky: There is a function to spin up the containers and stop them when all the tests using that fixture have finished. This function is wrapped in a different function that returns it. So I create a bunch of similarly named functions with the exact same functionality but a different scope, see here:
https://github.com/pytest-docker-compose/pytest-docker-compose/blob/4d505ae81afc2cfc557a51a29e4ce4e889f03e25/src/pytest_docker_compose/__init__.py#L58-L62

Being able to supply just one fixture and then let users determine it's scope dynamically would be a big improvement.

**Comment by RmStorm** [comment]
@nicoddemus quick question. Will the current implementation allow you to make two versions of the same fixture? One with module scope and one with function scope for example?

**Comment by nicoddemus** [comment]
Hi @RmStorm,

Thanks for the explanation, it definitely makes sense.

How about if you could control the scope of the fixture that spins up the docker containers via command-line, so you can for example have them spin up/down for every test in CI, and only once per session while in development? Does that make sense?

> Will the current implementation allow you to make two versions of the same fixture? One with module scope and one with function scope for example?

Unfortunately no, it is just a way to somehow change the scope of the fixture dynamically at definition time.

**Comment by RmStorm** [comment]
> Hi @RmStorm,
> 
> Thanks for the explanation, it definitely makes sense.
> 
> How about if you could control the scope of the fixture that spins up the docker containers via command-line, so you can for example have them spin up/down for every test in CI, and only once per session while in development? Does that make sense?
> 
> > Will the current implementation allow you to make two versions of the same fixture? One with module scope and one with function scope for example?
> 
> Unfortunately no, it is just a way to somehow change the scope of the fixture dynamically at definition time.

That's still pretty cool! And I think you can still use docker containers as an example but basically any fixture that deals with a large amount of data or is time consuming can benefit from dynamic scoping. At least that's how I see it.
